jwst/datamodels/history.py

```diff
@@ -0,0 +1,73 @@
+from collections import UserList
+
+from asdf import AsdfFile
+from asdf.tags.core import HistoryEntry
+
+class HistoryList:
+    """
+    A list that coerces a new value into a HistoryEntry.
+    Only a subset of the list interface is implemented.
+    """
+    def __init__(self, asdf):
+        self._context = asdf
+        if len(self._context.get_history_entries()):
+            self._entries = self._context.get_history_entries()
+        else:
+            self._context.add_history_entry("fake entry")
+            self._entries = self._context.get_history_entries()
+            self._entries.clear()
+
+    def __len__(self):
+        return len(self._entries)
+
+    def __getitem__(self, key):
+        return self._entries[key]
+
+    def __setitem__(self, key, value):
+        self.append(value)
+        value = self._entries.pop()
+        self._entries[key] = value
```

Comment:
The two parts of adding a new entry are constructing it and then validating it.  My code tries to avoid replicating these operations, so it won't get out of synch if the asdf code changes. The only operation for creating a new entry is add_history_entry, which append calls. So I append a new entry, pop the entry off, and write it to where the user wants to put it.